src/Types.jl

```diff
@@ -892,52 +893,251 @@ function ensure_resolved(env::EnvCache,
     pkgerror(msg)
 end
 
-const DEFAULT_REGISTRIES = Dict("General" => "https://github.com/JuliaRegistries/General.git")
+##############
+# Registries #
+##############
+
+mutable struct RegistrySpec
+    name::Union{String,Nothing}
+    uuid::Union{UUID,Nothing}
+    url::Union{String,Nothing}
+    # the path field can be a local source when adding a registry
+    # otherwise it is the path where the registry is installed
+    path::Union{String,Nothing}
+    RegistrySpec(name::String) = RegistrySpec(name = name)
+    RegistrySpec(;name=nothing, uuid=nothing, url=nothing, path=nothing) =
+        new(name, isa(uuid, String) ? UUID(uuid) : uuid, url, path)
+end
+
+const DEFAULT_REGISTRIES =
+    RegistrySpec[RegistrySpec(name = "General",
+                              uuid = UUID("23338594-aafe-5451-b93e-139f81909106"),
+                              url = "https://github.com/JuliaRegistries/General.git")]
 
-# Return paths of all registries in a depot
-function registries(depot::String)::Vector{String}
+# Return `RegistrySpec`s of each registry in a depot
+function collect_registries(depot::String)
     d = joinpath(depot, "registries")
-    ispath(d) || return String[]
-    regs = filter!(readdir(d)) do r
-        isfile(joinpath(d, r, "Registry.toml"))
+    regs = RegistrySpec[]
+    ispath(d) || return regs
+    for name in readdir(d)
+        for slug in readdir(joinpath(d, name))
```

Comment:
I think this is fine for now, but it will error out if `joinpath(d, name)` is not a directory right? I'm starting to believe we will need to validate depots sometime soon so we can avoid erroring out if assumptions like these don't hold (which shouldn't be the case, but it's easy for users to poke around and mess stuff up).